rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,278 @@
+"stage" which resources they want to request access to and store this state
+locally, and finally execute the full request without needing each UUID on the
+command line. But this may be too complex and confusing. With this MVP version
+users can copy-paste UUIDs from multiple searches, the meaning is clear, and it
+is fully customizable and scriptable.
+
+### On-Demand SSH
+
+Many times users would not want to search and request access in two steps.
+Teleport will have a new alias flag modifier that will try accessing a node, in
+case if access denied, it will request access for a node all in one step:
+
+```bash
+# -P "Please" activate try and request mode
+$ tsh ssh -P root@db-node
+# by default, if tsh will succeed to SSH, it will just continue with a
+# Session, otherwise it will find a node and create a search based access request
+You do not have access to the system by default, created access request.
+
+Please wait...
+
+Access request has been approved. Happy hacking!
+$
+```
+
+Some customers would want to have `-P` flag as a default. New alias feature in
+profile, will allow those users to set `tsh ssh` as an alias for `tsh ssh -P`.
+
+For OpenSSH use-cases, some users would like to load the new certificates in the
+agent, `tsh login -A root@db-node` will work like `tsh ssh -P` command except
+instead of login, it will load the keys in the agent.
+
+### Role Spec
+
+To enable this level of matching, we propose an extension to the role spec to
+determine the final roles that need to be granted.
+
+We will specify two roles that define this flow.
+
+The role `response-team` will allow user with this role to
+request access to nodes of teams “dbs” and “splunk”, and members of OIDC/SAML
+group “app” developers to request access to their own group.
+
+The spec for role-local variables will be expanded in a separate RFD (will post
+link here).
+
+```yaml
+kind: role
+metadata:
+  name: response-team
+spec:
+  # Local variables are necessary to specify additional rules
+  # that are not available in OIDC/SAML claims.
+  # Teleport admins don't always have control over OIDC/SAML.
+  # Specifying local variables allows them to add additional cases used in roles.
+  # Having role-local variables simplifies computation and role evaluation,
+  # as role eval keeps everything in one place.
+  vars:
+    - name: access_group
+      select: external.claims[groups]:
+        - "splunk": ["dbs", "{{external.claims['group']}}"]
+        # by default, allow members of the group to request access to their own group
+        - ".*": "{{external.claims['group']}}"
+  allow:
+    request:
+      # search_as_roles will allow a member of the response team
+      # to search resources that they will be allowed to request
+      # access to as a member of db-admin-role
+      # delegate details of the existing permissions to a special
+      # role that specifies that access.
+      # E.g the role can have only one logins: root parameter and that's all.
```

Comment:
These clauses both confused me. Perhaps they can be removed, or at least clarified.